### DISCUSSION TITLE
Mixed-model hierarchical content structures

### DISCUSSION THREAD

**Discussion by RobertAlbus** [payloadcms/payload#197] [Feature Requests & Ideas]
Hi! I really like Payload so far!

I am having a hard time finding any CMS (outside of a few enterprise options such as Sitecore) that allows me to **both** create custom data models **and** organize my content as a hierarchical tree. What I really like with Payload is that I can create the data model with freakin TypeScript!

With Payload it's clear I can set up the child-based hierarchical relationship with a RelationshipField. What I would really love is 

- bidirectional relationships for easy parent-child traversal
- a hierarchical content editor instead of a bucket-based editor

This is really two feature requests but they are directly related. If there is a way for me to contribute this functionality I would certainly be willing. 

For reference I have included two screenshots from a CMS which has this functionality. 

Content tree
![image](https://user-images.githubusercontent.com/42300645/121631060-9d8bd400-ca3b-11eb-84e8-750d5d5ef245.png)

Context menu with "Insert Options" being user-definable
![image](https://user-images.githubusercontent.com/42300645/121634105-412bb300-ca41-11eb-9d45-a91fd6298f98.png)

**Comment by Ontopic** [comment]
+1

I was able to use Payload's freedom on top of its logical solid core to achieve something like you're describing that feels sustainable, but I'd feel more comfortable with a more guided solution in line with Payload's vision for the future. The complete website series and demo repos really are needed to fully understand what's possible and how to address issues, so be sure to check those out, although I feel a more generalised set of examples will truly speed up onboarding. I'm sure the guys working on Payload are working hard towards goals like that. An update from them on where this is headed and a rough indication on when to "count on this or that" would be very helpful - even if it's a milestone more than a year from now

If you think it would be any help I could share some raw code / structures for now I plan on at least cleaning up to submit as general examples in a pull request in the future, but for now are too messy. So send me a private message if you care, although I'll ask for your insights with hopefully some actual code in return as well ;)

Also looking forward to contributing wherever feasible

**Reply by RobertAlbus** [reply]
Thanks @Ontopic for the advice to check out the website series and demo repo. I will dig into this ASAP. In the mean time, please do share code. What's the best way to reach you? Forgive the anti-bot string interpolation but I am available at `hello@${myUsername}.com`.
  
I think Payload is promising - I have dockerized more CMS' than I can count last week... and Payload was an instant obvious outlier. Thanks for the suggestion and please do send me some code to check out.

**Reply by Ontopic** [reply]
Checking out the demo folder in the main repo and running it locally is also a great way to figure out how to approach certain issues. I'll prepare some interesting patterns I feel are worthwhile and share them

**Reply by DanRibbens** [reply]
I really appreciate the discussion here. You guys are inspiring us to continue to work hard. 

I'm intrigued by the top down admin UI approach. I haven't been exposed to that kind of CMS before but I can see how it would be useful to organize that way.

We'll consider this for our roadmap, coming soon!

**Reply by Ontopic** [reply]
https://github.com/frontend-collective/react-sortable-tree is perhaps useful :)

**Reply by Ontopic** [reply]
The slate plugins also secretly offer the most insightful view for a dev in "tree-mode". The way nodes are recursively structured with a `type` and `children` field.
